x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/step/info/AllocationInfo.java

```diff
@@ -0,0 +1,120 @@
+ */
+
+package org.elasticsearch.xpack.core.ilm.step.info;
+
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.xcontent.ConstructingObjectParser;
+import org.elasticsearch.common.xcontent.ToXContentObject;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * Represents the state of an index's shards allocation, including a user friendly message describing the current state.
+ * It allows to transfer the allocation information to {@link org.elasticsearch.common.xcontent.XContent} using
+ * {@link #toXContent(XContentBuilder, Params)}
+ */
+public class AllocationInfo implements ToXContentObject {
+
+    private final long numberOfReplicas;
+    private final long numberShardsLeftToAllocate;
+    private final boolean allShardsActive;
+    private final String message;
+
+    static final ParseField NUMBER_OF_REPLICAS = new ParseField("number_of_replicas");
+    static final ParseField SHARDS_TO_ALLOCATE = new ParseField("shards_left_to_allocate");
+    static final ParseField ALL_SHARDS_ACTIVE = new ParseField("all_shards_active");
+    static final ParseField MESSAGE = new ParseField("message");
+    static final ConstructingObjectParser<AllocationInfo, Void> PARSER = new ConstructingObjectParser<>("allocation_routed_step_info",
+        a -> new AllocationInfo((long) a[0], (long) a[1], (boolean) a[2], (String) a[3]));
+
+    static {
+        PARSER.declareLong(ConstructingObjectParser.constructorArg(), NUMBER_OF_REPLICAS);
+        PARSER.declareLong(ConstructingObjectParser.constructorArg(), SHARDS_TO_ALLOCATE);
+        PARSER.declareBoolean(ConstructingObjectParser.constructorArg(), ALL_SHARDS_ACTIVE);
+        PARSER.declareString(ConstructingObjectParser.constructorArg(), MESSAGE);
+    }
+
+    public AllocationInfo(long numberOfReplicas, long numberShardsLeftToAllocate, boolean allShardsActive, String message) {
+        this.numberOfReplicas = numberOfReplicas;
+        this.numberShardsLeftToAllocate = numberShardsLeftToAllocate;
+        this.allShardsActive = allShardsActive;
+        this.message = message;
+    }
+
+    /**
+     * Builds the AllocationInfo representing a cluster state with a routing table that does not have enough shards active for a
+     * particular index.
+     */
+    public static AllocationInfo waitingForActiveShardsAllocationInfo(long actualReplicas) {
+        return new AllocationInfo(actualReplicas, -1, false,
+            "Waiting for all shard copies to be active");
+    }
+
+    /**
+     * Builds the AllocationInfo representing a cluster state with a routing table that has all the shards active for a particular index
+     * but there are still {@link #numberShardsLeftToAllocate} left to be allocated.
+     */
+    public static AllocationInfo allShardsActiveAllocationInfo(long actualReplicas, long numberShardsLeftToAllocate) {
```

Comment:
Super minor, but it's confusing to call the parameter here (and in the function above) "`actualReplicas`", it makes it seem like there should be a "`allocatedReplicas`" parameter somewhere as well. Maybe we can just call in `numReplicas`?

Reply:
Totally, my grep skills failed me.